include/nbl/video/CThreadSafeGPUQueueAdapter.h

```diff
@@ -0,0 +1,49 @@
+#ifndef __IRR_GPU_QUEUE_THREADSAFE_ADAPTER_H_INCLUDED__
+#define __IRR_GPU_QUEUE_THREADSAFE_ADAPTER_H_INCLUDED__
+#include "IGPUQueue.h"
+
+namespace nbl::video
+{
+    class CThreadSafeGPUQueueAdapter : public IGPUQueue
+    {
+    protected:
+        nbl::core::smart_refctd_ptr<IGPUQueue> originalQueue = nullptr;
+        std::mutex m;
+    public:
+        CThreadSafeGPUQueueAdapter(nbl::core::smart_refctd_ptr<IGPUQueue>&& originalQueue) : IGPUQueue(nullptr, originalQueue->getFamilyIndex(), originalQueue->getFlags(), originalQueue->getPriority()), originalQueue(originalQueue) {}
+        CThreadSafeGPUQueueAdapter(CThreadSafeGPUQueueAdapter&& other) : IGPUQueue(nullptr, other.originalQueue->getFamilyIndex(), other.originalQueue->getFlags(), other.originalQueue->getPriority()), originalQueue(other.originalQueue) {}
+        
+        CThreadSafeGPUQueueAdapter(const CThreadSafeGPUQueueAdapter& other) : IGPUQueue(nullptr, other.originalQueue->getFamilyIndex(), other.originalQueue->getFlags(), other.originalQueue->getPriority()), originalQueue(other.originalQueue) {}
+        CThreadSafeGPUQueueAdapter(const nbl::core::smart_refctd_ptr<IGPUQueue>& originalQueue) : IGPUQueue(nullptr, originalQueue->getFamilyIndex(), originalQueue->getFlags(), originalQueue->getPriority()), originalQueue(originalQueue) {}
+        
+        CThreadSafeGPUQueueAdapter() : IGPUQueue(nullptr, 0, E_CREATE_FLAGS::ECF_PROTECTED_BIT, 0.f) {};
```

Comment:
this makes no sense, `IGPUQueue` is a "Total Interface"

also, as a convention you should never make a function that takes a `const core::smart_refctd_ptr<T>&` if you take a smartpointer yuo shall always take `core::smart_refctd_ptr<T>&&`

Reply:
in general, we dont waste time on writing all ctors for refcounted classes, they'll never be used anyway, make it inherit from core::Uncopyable and core::Unmovable is best what you can do

Reply:
`IGPUQueue` already does inherit bboth from Unmovable and Uncopyable, but the problem is that if you declare your own copy ctors and assignment operators on the derived class, there's nothing the parent can do about that

Reply:
resolved